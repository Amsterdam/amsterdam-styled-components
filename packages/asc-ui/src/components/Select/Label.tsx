import React from 'react'
import { AscCore } from '@datapunt/asc-core'
import { ReactComponent as Elipsis } from '@datapunt/asc-assets/lib/Icons/Elipsis.svg'
import { ReactComponent as ChevronTop } from '@datapunt/asc-assets/lib/Icons/Chevron-Top.svg'
import { ReactComponent as ChevronDown } from '@datapunt/asc-assets/lib/Icons/Chevron-Down.svg'
import { Icon } from '../..'

const Label = ({ id, label, open, onOpenClose }: any) => {
  return (
    <AscCore.Select.Label.default
      {...{ id }}
      onBlur={() => onOpenClose(false)}
      onClick={() => onOpenClose()}
    >
      {label || (
        <Icon inline>
          <Elipsis />
        </Icon>
      )}
      <Icon inline>{open ? <ChevronDown /> : <ChevronTop />}</Icon>
    </AscCore.Select.Label.default>
  )
}

export default Label
