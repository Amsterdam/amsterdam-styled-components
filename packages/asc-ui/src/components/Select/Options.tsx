import React from 'react'
import { AscCore } from '@datapunt/asc-core'

type Props = {
  children: React.ReactChildren & {
    props: AscCore.Select.Option.Props
  }
  currentTab: string
}

const Options = ({ children: childrenProp, currentTab }: Props) => {
  let childIndex = 0
  const children = React.Children.map(childrenProp, child => {
    if (!React.isValidElement(child)) {
      return null
    }

    const { label } = child.props
    const childValue = label === undefined ? childIndex : label
    const selected = childValue === currentTab

    childIndex += 1
    return React.cloneElement(child as React.ReactElement<any>, {
      isCurrentTab: selected,
    })
  })
  return <>{children}</>
}

export default Options
