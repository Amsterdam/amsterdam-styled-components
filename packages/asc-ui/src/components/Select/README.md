# Select

[Design System](https://designsystem.amsterdam.nl/7awj1hc9f/p/6561a5-selection-dropdown)

**Component status: experimental**
