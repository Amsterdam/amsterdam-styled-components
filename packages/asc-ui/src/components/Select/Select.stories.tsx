import React from 'react'
import { storiesOf } from '@storybook/react'
import { Select, Option } from '../..'

storiesOf('Selection', module).add('Default', () => (
  <Select small id="hello">
    <Option.default id="listbox1-1" className="selected" aria-selected="true">
      Green
    </Option.default>
    <Option.default id="listbox1-2">Orange</Option.default>
  </Select>
))
