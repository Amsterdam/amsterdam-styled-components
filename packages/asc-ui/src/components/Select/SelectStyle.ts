import { ChevronDown } from '@amsterdam/asc-assets'
import { CSSProperties } from 'react'
import styled, { css } from 'styled-components'
import { themeColor, themeSpacing } from '../../utils'
import { INPUT_SIZE } from '../shared/constants'

export type Props = {
  id?: string
  value?: string
  label?: string
  disabled?: boolean
  srOnly?: boolean
  error?: boolean | string
  errorMessage?: string
  labelStyle?: CSSProperties
  errorStyle?: CSSProperties
}

export const SelectWrapper = styled.div`
  position: relative;
  height: ${INPUT_SIZE}px;
  width: 100%;
`

/**
 * Contains all of the absolute content which is placed above the select element.
 */
export const AbsoluteContentWrapper = styled.div.attrs({
  'aria-hidden': 'true',
})`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 3px; /* Allows the default browser(s) default focus style to be displayed. */
  padding: 0 calc(${themeSpacing(3)} - 1px); /* Match the spacing of the select element. */
  display: flex;
  align-items: center;
  pointer-events: none;
  background-color: ${themeColor('tint', 'level1')};
`

/**
 * Since Chrome does not support `text-overflow: hidden` we need show the value in a separate element.
 */
export const SelectedValue = styled.div<{ disabled?: boolean }>`
  margin-right: auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
    `}
`
export const SelectIcon = styled(ChevronDown)`
  display: block;
  width: ${themeSpacing(3)};
  height: ${themeSpacing(3)};
  margin-left: ${themeSpacing(3)};
`

const SelectStyle = styled.select<Props>`
  width: 100%;
  height: 100%;
  padding: ${themeSpacing(2, 3)};
  font-size: 1rem;
  border: 1px solid ${themeColor('tint', 'level5')};
  border-radius: 0;
  background-color: ${themeColor('tint', 'level1')};
  appearance: none;
  cursor: pointer;

  /* IE11 (hide native arrow button) */
  &::-ms-expand {
    display: none;
  }

  /* IE11 (hide native focus state) */
  &::-ms-value {
    color: black;
    background: transparent;
  }

  /* Firefox (hide native focus state) */
  &:-moz-focusring {
    color: transparent;
    text-shadow: 0 0 0 ${themeColor('tint', 'level7')};
  }

  &:hover {
    ${({ disabled, error }) =>
      !disabled &&
      !error &&
      css`
        border: 1px solid ${themeColor('tint', 'level7')};
      `}
  }

  &:disabled {
    pointer-events: none;
    opacity: 0.5;
    border: 1px solid ${themeColor('tint', 'level4')};
  }

  ${({ error }) =>
    error &&
    css`
      border: 1px solid ${themeColor('error', 'main')};
    `}
`

export default SelectStyle
