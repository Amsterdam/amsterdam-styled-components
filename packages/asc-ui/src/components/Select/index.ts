export { default as SelectStyle } from './SelectStyle'
export { default } from './Select'
