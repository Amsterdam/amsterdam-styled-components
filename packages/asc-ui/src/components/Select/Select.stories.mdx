import { createRef, useEffect } from 'react'
import { Meta, Story, Preview } from '@storybook/addon-docs/blocks'
import Select from './Select'
import ErrorMessage from '../ErrorMessage'

export const SelectWithRef = () => {
  const ref = createRef()
  useEffect(() => {
    const element = ref.current
    function onChange(e) {
      const { value } = e.target
      // eslint-disable-next-line no-console
      console.log('Value changed!', value)
    }
    if (!element) {
      return undefined
    }
    element.addEventListener('change', onChange)
    return () => {
      element.removeEventListener('change', onChange)
    }
  }, [ref])
  return (
    <Select id="with-ref" label="Select with ref" ref={ref}>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  )
}

<Meta
  title="Experimental/Atoms/Select (Dropdown)"
  component={Select}
  decorators={[
    (storyFn) => (
      <div
        style={{
          width: '400px',
          padding: '40px 10px',
        }}
      >
        {storyFn()}
      </div>
    ),
  ]}
/>

# Select (dropdown)

Use this component as a last resort, because some users are having trouble
with it. Try to use Radio Buttons first, or in case of a long list, try
to ask multiple questions (text from Design System)

[Design System](https://designsystem.amsterdam.nl/7awj1hc9f/p/6561a5-selection-dropdown)

**Component status: experimental**

## Default state

<Preview>
  <Story name="Default state">
    <Select id="select-1" label="Label">
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </Story>
</Preview>

## Selected state

<Preview>
  <Story name="Selected state">
    <Select id="select-1" label="Label">
      <option value="1">Option 1</option>
      <option value="2" selected>
        Option 2
      </option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </Story>
</Preview>

## Disabled state

<Preview>
  <Story name="Disabled state">
    <React.Fragment>
      <Select id="select-1" label="Disabled Select component" disabled>
        <option value="1">Option 1</option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
      <br />
      <br />
      <Select id="select-2" label="Disabled specific Option">
        <option value="1" disabled>
          Option 1
        </option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
    </React.Fragment>
  </Story>
</Preview>

## Error state

<Preview>
  <Story name="Error state">
    <React.Fragment>
      <Select id="select-1" label="With Error Message" error>
        <option value="1">Option 1</option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
      <ErrorMessage message="Error message" />
      <br />
      <br />
      <Select id="select-2" label="Without Error Message" error>
        <option value="1">Option 1</option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
    </React.Fragment>
  </Story>
</Preview>

## With ref

<Preview>
  <Story name="With ref">
    <SelectWithRef />
  </Story>
</Preview>
