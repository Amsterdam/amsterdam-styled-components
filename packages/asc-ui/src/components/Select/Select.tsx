import React from 'react'
import { AscCore } from '@datapunt/asc-core'
import Label from './Label'

type Props = {
  id: string
  orientation?: 'top'
  label?: string
  small?: boolean
}

type State = {}

class Selection extends React.Component<Props, State> {
  state = {
    open: false,
  }

  onOpenClose = (override?: boolean) => {
    const { open } = this.state
    this.setState({
      open: override || !open,
    })
  }

  render() {
    const { id, label, children, orientation }: any = this.props
    const { open } = this.state
    return (
      <AscCore.Select.Wrapper.default>
        <Label
          {...{
            open,
            id,
            label,
          }}
          onOpenClose={this.onOpenClose}
        />
        <AscCore.Select.ListBox.default
          labelId={id}
          orientation={orientation}
          aria-hidden={!open}
          aria-activedescendant="listbox1-1"
        >
          {children}
        </AscCore.Select.ListBox.default>
      </AscCore.Select.Wrapper.default>
    )
  }
}

export default Selection
