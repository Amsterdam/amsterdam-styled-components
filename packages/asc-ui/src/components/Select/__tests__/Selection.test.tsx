import React from 'react'
import { Option } from '@datapunt/asc-ui/'
import { shallow } from 'enzyme'
import Select from '../Select'

describe('Selection', () => {
  const component = shallow<Select>(
    <Select small id="foo" label="Click on me">
      <Option.default id="listbox1-1" className="selected" aria-selected="true">
        Green
      </Option.default>
      <Option.default id="listbox1-2">Orange</Option.default>
    </Select>,
  )

  it('should render', () => {
    expect(component).toMatchSnapshot()
  })

  describe('state', () => {
    const instance = component.instance()
    const label = component.find('#foo').at(0)

    it('should toggle the open state', () => {
      label.simulate('click')
      expect(instance.state.open).toBe(true)

      label.simulate('click')
      expect(instance.state.open).toBe(false)
    })

    it('should set the open state to false', () => {
      instance.state.open = true

      label.simulate('blur')
      expect(instance.state.open).toBe(false)
    })
  })
})
