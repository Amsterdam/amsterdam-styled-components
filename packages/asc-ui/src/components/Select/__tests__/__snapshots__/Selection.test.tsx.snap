// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection should render 1`] = `
<Fragment>
  <styled.button
    id="foo"
    onBlur={[Function]}
    onClick={[Function]}
  >
    <span>
      Click on me
    </span>
    <styled.span
      inline={true}
    >
      <Component />
    </styled.span>
  </styled.button>
  <styled.div
    aria-activedescendant="listbox1-1"
    aria-hidden={true}
    id="foo"
  >
    <styled.button
      aria-selected="true"
      className="selected"
      id="listbox1-1"
    >
      Green
    </styled.button>
    <styled.button
      id="listbox1-2"
    >
      Orange
    </styled.button>
  </styled.div>
</Fragment>
`;
